### ISSUE TITLE
App idea: Password Manager

### ISSUE THREAD

**Issue by dpc** [urbit/urbit#1219]
Please excuse me putting this here, but right now my start is disconnected again, fora is down, reddit is obsolete, and there's basically no public place to discuss urbit ideas.

Anyway,

I was reviewing stuff like SuperGenPass, LessPass, Master Password (deterministic password managers), and it occurred to me that Urbit could be really useful for making these mechanism much more secure.

To recap, deterministic password manager takes (*domain*, *master password*, *potentially other stuff*) and generates per-website passwords. The biggest weakness of this approach is - in case the master password is compromised, all passwords derived from it are.

It seems to me it would be fairly easy to supplement this approach with Urbit, to eliminate the weakest point.

A deterministic password manager would do what it normally does, and in addition, it would request additional password from urbit: be sending it the same (*domain* and *potentially other stuff*) to https://ship-name.urbit.org/ and receiving in response additional deterministic entropy, derived from a different *master password*, stored on Urbit. These two can be simply combined (concat + sha256 or whatever).

This approach makes it so attacker would have to obtain both master keys: the one provided by the user and the one stored on urbit.

Each app would authenticate with urbit through some `+code`-like mechanism, and rate-limiting could be implemented. Once Urbit runs moons on mobile phones, it could be used for asking user to confirm each password request.

The beauty of this approach is, that "urbit salt" could be easily implemented as an additional feature/fork of popular password managers (LessPass, Master Password), so a lot of UI work would be already done.

**Comment by gtasteve** [comment]
There is an unofficial urbit discord group where we are sharing info:
 https://discord.gg/S79CqQ4

**Comment by msutherl** [comment]
@dpc we're designing a new proposal system at the moment. I'll make sure this ends up there.

**Comment by galenwp** [comment]
We've kicked this idea around a bit (a long time ago). It's nice to think about — but I'd be scared to have people putting secrets into Arvo without a real audit. And that's quite a challenging task.

I wouldn't be averse to seeing work on the application logic / interface. So long as it comes with lots of Big Red Warnings about Do Not Use This For Real etc.

**Comment by philipcmonk** [comment]
I like this proposal because compromising arvo would only reduce the security to the level of the existing best practices.

**Comment by dpc** [comment]
> I like this proposal because compromising arvo would only reduce the security to the level of the existing best practices.

Exactly. As long as it's being used to supplement an otherwise entirely local password generation, Urbit works just as a 2nd factor. If it gets compromised, the user just reverts to security provided by deterministic password generator.

I was thinking that I would like to use deterministic password manager, but by itself it has weaknesses. I'd like 2nd factor on that. But I need something that is publicly reachable and reasonably secure. And if it's 3rd party then I compromise on privacy. So I would like to run it myself. That might work for me, because I am skilled Unix professional, but not for "normal users". So then I'm thinking ... what is supposed to be privately owned, maintenance free, can have a built-in human interface, can do crypto and is publicly reachable? Urbit. This is exactly the sort of application that fits "small notebook for your small data". And something people would be compelled to try, pay $10 for an account etc. Just for this single app! If it's easy enough.

And this is generalizable.
